lib/galaxy/objectstore/irods.py

```diff
@@ -662,6 +662,7 @@ def _update_from_file(self, obj, file_name=None, create=False, **kwargs):
                         # FIXME? Should this be a `move`?
                         shutil.copy2(source_file, cache_file)
                     self._fix_permissions(cache_file)
+                    source_file = cache_file
```

Comment:
This is copy-pasted from the S3 object store and possibly others, so if this is the correct fix it should be applied everywhere (and this needs to be refactored to a common base ASAP).

Reply:
Alright, so what is happening is that the definition of put in python-irods is
```
    def put(self, local_path, irods_path, return_data_object = False, num_threads = DEFAULT_NUMBER_OF_THREADS, **options):

        if irods_path.endswith('/'):
            obj = irods_path + os.path.basename(local_path)
        else:
            obj = irods_path
```
and we're calling this here:

```
self.session.data_objects.put(data_obj.desc, f"{collection_path}/", **options)
```

The proposed fix works because the `cache_file`'s basename has the correct pattern, as opposed to file_name which can be anything. But that means we're copying the file from the potentially remote cache location, instead of the possibly local file. I think the following patch should work:
```
diff --git a/lib/galaxy/objectstore/irods.py b/lib/galaxy/objectstore/irods.py
index eb6841f6c5..9b9bfa4bf5 100644
--- a/lib/galaxy/objectstore/irods.py
+++ b/lib/galaxy/objectstore/irods.py
@@ -429,7 +429,7 @@ class IRODSObjectStore(DiskObjectStore, CloudConfigMixin):
                 log.debug("Pushing cache file '%s' of size %s bytes to collection '%s'", source_file, os.path.getsize(source_file), rel_path)
 
                 # Add the source file to the irods collection
-                self.session.data_objects.put(source_file, f"{collection_path}/", **options)
+                self.session.data_objects.put(source_file, data_object_path, **options)
 
                 end_time = datetime.now()
                 log.debug("Pushed cache file '%s' to collection '%s' (%s bytes transfered in %s sec)",

```

@kxk302 can you add a test that verifies that file uploads are working correctly also when the cache region is deleted ? You could take some inspiration from this test: https://github.com/mvdbeek/galaxy/blob/11c5a1923ee944a4cb558660a0d2e624e4bb9b39/test/integration/objectstore/test_remote_objectstore_cache_operations.py#L28

Reply:
This same call was working fine for months. Somebody made a change upstream and broke it. Apparently, it was not tested.

Reply:
@mvdbeek Can you please clarify the purpose of the test that you suggested?

Reply:
A regression test that will use `rel_path`. This integration test (if you choose to do integration) will fail prior to the fix because it only populates the local cache, but puts the file into a collection called `<uuid>.dat` with the basename of `rel_path`, and that path won't be retrievable.

Reply:
@mvdbeek There are roughly 3 steps in an upload job:

1) Pre-processing: a file of size is created in the cache and pushed to irods
2) The upload job runs and a file is created in the working directory
3) Post-processing: File created in the working directory in step 2 is copied to cache, and pushed to irods 

My understanding is that in your proposed test, cache is cleared before pushing a file to irods in Step 3? Is that correct?

Reply:
@mvdbeek any updates on my question? Thx.

Reply:
Probably ? Make sure it's a regression test that fails before the fix.